File: src/loop.ts
import { FailureContext, Hypothesis, HypothesisStatus, ConfidenceLevel, Evidence, EvidenceSourceType, Diagnosis, ProgressCallback } from './domain';
import { ILLMProvider } from './llm';
import { ExperimentRunner } from './execution';
import { HypothesisGenerator } from './generator';
import { Ledger } from './ledger';
import { Verifier } from './verifier';
import { Patcher } from './patcher';
import { Observability } from './observability';

export class InvestigationLoop {
  constructor(
    private llm: ILLMProvider,
    private runner: ExperimentRunner,
    private ledger: Ledger,
    private onProgress?: ProgressCallback
  ) {}

  private emit(msg: string) {
    if (this.onProgress) this.onProgress(msg);
  }

  async run(context: FailureContext): Promise<Diagnosis> {
    const obs = Observability.getInstance();
    const sessionId = `session-${Date.now()}`;
    obs.startSession(sessionId, this.runner.getTargetRepoPath());

    const generator = new HypothesisGenerator(this.llm);
    const verifier = new Verifier(this.llm);
    const patcher = new Patcher(this.llm);
    
    this.emit('Generating initial hypotheses from failure context...');
    let hypotheses = await generator.generate(context);
    
    for (const h of hypotheses) {
      this.ledger.addHypothesis(h);
      this.emit(`-> Hypothesis created: ${h.statement.substring(0, 50)}...`);
    }

    let loopCount = 0;
    const sleep = (ms: number) => new Promise(r => setTimeout(r, ms));
    
    while (loopCount < 8) { 
      loopCount++;
      const pendingHypothesis = this.ledger.getAllHypotheses().find(h => h.status === HypothesisStatus.PENDING);
      
      if (!pendingHypothesis) break;

      this.emit(`\n[Step {loopCount}] Investigating: ${pendingHypothesis.statement.substring(0, 50)}...`);
      const prompt = `Hypothesis: ${pendingHypothesis.statement}\nContext: ${JSON.stringify(context)}\nPropose next experiment command. Options: read file <path>, search files <query>, run pytest, inspect git diff. Return ONLY the command string.`;
      
      if (process.env.NODE_ENV !== 'test') {
        this.emit('Spacing request to respect Gemini rate limits...');
        await sleep(7000);
      }

      let command = await this.llm.generate({ 
        systemPrompt: 'You are an investigator.', 
        userPrompt: prompt,
        promptName: 'AgentOrchestrator_NextCmd'
      });
      command = command.trim().replace(/^`+|`+$/g, '');

      if (!command) {
        pendingHypothesis.status = HypothesisStatus.INCONCLUSIVE;
        continue;
      }

      this.emit(`-> Running experiment: ${command}`);
      let result;
      try {
        result = await this.runner.runOperation(command);
      } catch (e: any) {
        this.emit(`-> Experiment failed: ${e.message}`);
        pendingHypothesis.status = HypothesisStatus.INCONCLUSIVE;
        continue; 
      }

      this.emit(`-> Evaluating experiment evidence...`);
      const evalPrompt = `Command: ${command}\nOutput: ${result.output}\nDoes this support or contradict the hypothesis: "${pendingHypothesis.statement}"? Return JSON: {"supports": true/false, "contradicts": true/false, "reason": "..."}`;
      
      if (process.env.NODE_ENV !== 'test') {
        await sleep(7000);
      }

      let evalResultStr = await this.llm.generate({ 
        systemPrompt: 'You evaluate evidence strictly.', 
        userPrompt: evalPrompt,
        promptName: 'AgentOrchestrator_EvalEvidence'
      });
      
      let supports = false;
      let contradicts = false;
      try {
        const cleaned = evalResultStr.trim().replace(/^```json/, '').replace(/```$/, '').trim();
        const evEval = JSON.parse(cleaned);
        supports = evEval.supports;
        contradicts = evEval.contradicts;
      } catch (e) {}

      const evidence: Evidence = {
        id: `ev-${Date.now()}-${loopCount}`,
        hypothesisId: pendingHypothesis.id,
        sourceType: EvidenceSourceType.COMMAND_OUTPUT,
        source: command,
        content: result.output.substring(0, 200),
        experimentId: `exp-${Date.now()}`,
        supports,
        contradicts
      };

      this.ledger.addEvidence(evidence);
      pendingHypothesis.status = this.ledger.evaluateHypothesisStatus(pendingHypothesis.id);
      
      if (pendingHypothesis.status === HypothesisStatus.SUPPORTED) {
        this.emit(`-> Hypothesis conditionally supported. Handing to Independent Verifier...`);
        const evFor = pendingHypothesis.evidenceFor.map(id => this.ledger.getEvidence(id)!);
        const evAgainst = pendingHypothesis.evidenceAgainst.map(id => this.ledger.getEvidence(id)!);
        if (process.env.NODE_ENV !== 'test') await sleep(7000);
        const verifiedStatus = await verifier.verify(pendingHypothesis, evFor, evAgainst, context);
        pendingHypothesis.status = verifiedStatus;
        this.emit(`-> Verifier verdict: ${verifiedStatus}`);
      }
    }

    const allHyp = this.ledger.getAllHypotheses();
    const supported = allHyp.find(h => h.status === HypothesisStatus.SUPPORTED);
    
    this.emit(`\nInvestigation Complete. Root cause found: ${supported ? 'YES' : 'NO'}`);
    
    const diagnosis: Diagnosis = {
      rootCause: supported ? supported.statement : 'Unknown',
      supportingEvidence: supported ? supported.evidenceFor : [],
      rejectedHypotheses: allHyp.filter(h => h.status === HypothesisStatus.REJECTED).map(h => h.id),
      experiments: this.runner.getRecords().map(r => r.command),
      confidence: supported ? ConfidenceLevel.HIGH : ConfidenceLevel.LOW,
      unresolvedQuestions: []
    };

    if (supported) {
      if (process.env.NODE_ENV !== 'test') await sleep(7000);
      const fix = await patcher.generatePatch(diagnosis, context);
      if (fix) diagnosis.suggestedFix = fix;
    }

    obs.endSession();
    return diagnosis;
  }
}

File: python/loop.py
import json
from typing import Callable, Optional
from .domain import FailureContext, Diagnosis, Evidence, HypothesisStatus
from .llm import ILLMProvider
from .execution import ExperimentRunner
from .ledger import Ledger
from .generator import HypothesisGenerator
from .verifier import IndependentVerifier
from .patcher import Patcher

ProgressCallback = Callable[[str], None]

class InvestigationLoop:
    def __init__(
        self,
        llm: ILLMProvider,
        runner: ExperimentRunner,
        ledger: Ledger,
        on_progress: Optional[ProgressCallback] = None
    ):
        self.llm = llm
        self.runner = runner
        self.ledger = ledger
        self.on_progress = on_progress
        self.generator = HypothesisGenerator(llm)
        self.verifier = IndependentVerifier(llm)
        self.patcher = Patcher(llm)

    def _emit(self, msg: str):
        if self.on_progress:
            self.on_progress(msg)

    def run(self, context: FailureContext) -> Diagnosis:
        self._emit("Generating initial hypotheses from failure context...")
        hypotheses = self.generator.generate_hypotheses(context)

        for h in hypotheses:
            self.ledger.add_hypothesis(h)
            self._emit(f"-> Hypothesis created: {h.statement[:65]}...")

        successful_hypothesis = None
        experiments_run = []

        for step, h in enumerate(hypotheses, 1):
            self._emit(f"\n[Step {step}] Investigating: {h.statement[:65]}...")
            command = h.proposed_experiment
            experiments_run.append(command)

            self._emit(f"-> Running experiment: {command}")
            result = self.runner.run_experiment(command)

            if not result.get("success", False):
                self._emit(f"-> Experiment failed: {result.get('output', '')}")
                continue

            output = result.get("output", "")
            self._emit("-> Evaluating experiment evidence...")

            # Analyze evidence with LLM
            eval_prompt = f"""
Analyze this command output to see if it supports or contradicts the hypothesis.

Hypothesis: {h.statement}
Command: {command}
Output:
{output[:1500]}

Respond ONLY in JSON:
{{
  "supports": boolean,
  "contradicts": boolean,
  "reason": string
}}
"""
            eval_resp = self.llm.generate(
                system_prompt="Analyze code evidence objectively. Respond in JSON.",
                user_prompt=eval_prompt
            )

            supports = False
            contradicts = False
            reason = "Analyzed output"

            try:
                cleaned = eval_resp.strip().replace("```json", "").replace("```", "").strip()
                parsed = json.loads(cleaned)
                supports = bool(parsed.get("supports", False))
                contradicts = bool(parsed.get("contradicts", False))
                reason = parsed.get("reason", "")
            except Exception:
                supports = True

            ev = Evidence(
                id=f"ev_{step}",
                hypothesis_id=h.id,
                source_type="FILE_READ" if "read file" in command or "cat" in command else "TEST_OUTPUT",
                source=command,
                content=output[:1000],
                supports=supports,
                contradicts=contradicts,
                reasoning=reason
            )
            self.ledger.add_evidence(ev)

            status = self.ledger.evaluate_hypothesis_status(h.id)
            if status == HypothesisStatus.SUPPORTED:
                self._emit("-> Hypothesis conditionally supported. Handing to Independent Verifier...")
                verdict = self.verifier.verify_evidence(h, ev)
                h.status = verdict
                self._emit(f"-> Verifier verdict: {verdict.value}")

                if verdict == HypothesisStatus.SUPPORTED:
                    successful_hypothesis = h
                    break
            else:
                h.status = status
                self._emit(f"-> Ledger evaluated status: {status.value}")

        self._emit(f"\nInvestigation Complete. Root cause found: {'YES' if successful_hypothesis else 'NO'}")

        if successful_hypothesis:
            diag = Diagnosis(
                root_cause=successful_hypothesis.statement,
                confidence="HIGH",
                experiments=experiments_run
            )
        else:
            diag = Diagnosis(
                root_cause="Inconclusive: multiple hypotheses refuted or insufficient repository evidence.",
                confidence="LOW",
                experiments=experiments_run
            )

        if diag.confidence != "LOW":
            self._emit("Synthesizing unified repair patch...")
            diag.suggested_fix = self.patcher.generate_patch(diag, context)

        return diag
